### ISSUE TITLE
rk4 sets number of rhs to 0

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#1079]
rk4 is very slow, roughly a factor of 10 compared to pvode at the same number of rhs evaluation.

At least to me this is surprising, as the rk4 scheme should not be that complicated compared to an rhs evaluation ...
It claims that 93% of the time is spent in Calc, only ~4.4 % in solver - which I doubt.

Is there something I am missing that can explain this?

**Comment by d7919** [comment]
Is the nrhs very different? Are you using it adaptively? The rk schemes do typically spend a short amount of time in the solver compared to the implicit solvers.

**Comment by d7919** [comment]
As the explicit solvers are limited by the CFL condition, whilst the implicit solvers are not limited in step size but instead just seeking a particular tolerance, the explicit solvers can often be fastest for low resolution cases but much slower for high resolution runs, at least in my experience.

**Comment by ZedThree** [comment]
Can you check rk4 vs rkgeneric and/or PETSc's rk solver? @d7919 what's the default order of rkgeneric? For PETSc, I think you also need to set `ts_rk_type` to 4, otherwise it uses a third order scheme.

**Comment by d7919** [comment]
There isn't a default order for rkgeneric -- you have to specify which scheme you want to use (probably something worth improving on by setting a sensible default).

**Comment by dschwoerer** [comment]
I tried with the blob2d example, where everything is fine.

I tested with the monitor example, and the issue is that the nrhs displayed are wrong, which explains the factor of 10 difference.

rkgeneric seems to be suffering from the same issue, i.e. resetting the number of rhs after calling the monitors, while they should be reset from the monitor that outputs the information.
Default for rkgeneric seems to be rkf45.
